cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup_test.go

```diff
@@ -0,0 +1,697 @@
+		testNodeGroupIncreaseMachineDeployment(t, i, tc)
+	}
+}
+
+func TestNodeGroupDecreaseSize(t *testing.T) {
+	for i, tc := range []testNodeGroupResizeTestCase{{
+		description: "errors because delta is positive",
+		delta:       1,
+		errors:      true,
+		initial:     3,
+		maxSize:     "10",
+		minSize:     "1",
+	}, {
+		description: "errors because delta exceeds node count",
+		delta:       -4,
+		errors:      true,
+		initial:     3,
+		maxSize:     "10",
+		minSize:     "1",
+	}, {
+		description: "no error as within bounds",
+		delta:       -1,
+		expected:    2,
+		initial:     3,
+		maxSize:     "10",
+		minSize:     "1",
+	}} {
+		t.Logf("test #%d: %s", i, tc.description)
+
+		testNodeGroupDecreaseMachineSet(t, i, tc)
+		testNodeGroupDecreaseMachineDeployment(t, i, tc)
+	}
+}
+
+func TestNodeGroupMachineSetDeleteNodes(t *testing.T) {
+	nodes := make([]*apiv1.Node, 10)
+	machines := make([]*v1beta1.Machine, 10)
+	nodeObjects := make([]runtime.Object, 10)
+	machineObjects := make([]runtime.Object, 10)
+
+	machineSet := &v1beta1.MachineSet{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineSet",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machineset",
+			Namespace: "test-namespace",
+			UID:       "abcdef12-a3d5-a45f-887b-6b49aa8fc218",
+		},
+		Spec: v1beta1.MachineSetSpec{
+			Replicas: int32ptr(int32(len(machines))),
+		},
+	}
+
+	for i := 0; i < len(nodes); i++ {
+		nodes[i] = &apiv1.Node{
+			TypeMeta: v1.TypeMeta{
+				Kind: "Node",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name: fmt.Sprintf("node-%d", i),
+				Annotations: map[string]string{
+					machineAnnotationKey: fmt.Sprintf("test-namespace/machine-%d", i),
+				},
+			},
+			Spec: apiv1.NodeSpec{
+				ProviderID: fmt.Sprintf("providerid-%d", i),
+			},
+		}
+
+		machines[i] = &v1beta1.Machine{
+			TypeMeta: v1.TypeMeta{
+				Kind: "Machine",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:      fmt.Sprintf("machine-%d", i),
+				Namespace: "test-namespace",
+				OwnerReferences: []v1.OwnerReference{{
+					Name: machineSet.Name,
+					Kind: machineSet.Kind,
+					UID:  machineSet.UID,
+				}},
+			},
+			Status: v1beta1.MachineStatus{
+				NodeRef: &apiv1.ObjectReference{
+					Kind: nodes[i].Kind,
+					Name: nodes[i].Name,
+				},
+			},
+		}
+
+		nodeObjects[i] = nodes[i]
+		machineObjects[i] = machines[i]
+	}
+
+	controller, stop := mustCreateTestController(t, testControllerConfig{
+		nodeObjects:    nodeObjects,
+		machineObjects: append(machineObjects, machineSet),
+	})
+	defer stop()
+
+	ng, err := newNodegroupFromMachineSet(controller, machineSet)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	nodeNames, err := ng.Nodes()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if len(nodeNames) != len(nodes) {
+		t.Fatalf("expected len=%v, got len=%v", len(nodes), len(nodeNames))
+	}
+
+	sort.Strings(nodeNames)
+
+	for i := 0; i < len(nodes); i++ {
+		if nodeNames[i] != nodes[i].Spec.ProviderID {
```

Comment:
This will work as long as `len(nodes)` < 10. With 10 or more you will get:
- `node-10`
- `node-11`
So sorting the list will yield `node-0, node-1, node-10, node-11, node2, ...` which is not what is expected here.

Reply:
This is an astute observation!

Reply:
Addressed in https://github.com/openshift/kubernetes-autoscaler/pull/34/commits/25a271f4c3981339dce51278de4b3e02b8efe8ef